packages/unmock-core/src/service/state/state.ts

```diff
@@ -111,50 +102,63 @@ export class State {
+        const nA = a.split("*");
+        const nB = b.split("*");
+        return nA > nB || (nA === nB && a.indexOf("*") < b.indexOf("*"))
+          ? -1
+          : 1;
+      })[0];
+    if (mostRelevantEndpoint === undefined) {
       debugLog(`No states match '${endpoint}'`);
       return undefined;
     }
-    // sort endpoints by location of asterisks and frequency
-    // (done to spread and overwrite as needed). Results is e.g:
-    // (**, /stores/*/petId/*, /stores/*/petId/404, /stores/myStore/petId/*, /stores/foo/petId/200)
-    matchingEndpointKeys.sort((a: string, b: string) => {
-      const nA = a.split("*");
-      const nB = b.split("*");
-      return nA > nB || (nA === nB && a.indexOf("*") < b.indexOf("*")) ? 1 : -1;
-    });
 
-    debugLog(
-      `Found following endpoints as matches for '${matchingEndpointKeys}: ${JSON.stringify(
-        matchingEndpointKeys,
-      )}`,
-    );
+    debugLog(`Most relevant endpoint is ${mostRelevantEndpoint}`);
 
-    // From the above, sorted methods, get all states that match either DEFAULT_STATE_HTTP_METHOD
-    // or the given method. Push the DEFAULT_STATE one before to maintain order.
-    const states: codeToMedia[] = [];
-    for (const key of matchingEndpointKeys) {
-      const methodToStatus = this.state[key];
-      if (methodToStatus[DEFAULT_STATE_HTTP_METHOD] !== undefined) {
-        states.push(methodToStatus[DEFAULT_STATE_HTTP_METHOD]);
-      }
-      if (methodToStatus[method] !== undefined) {
-        states.push(methodToStatus[method]);
-      }
-    }
+    // From the chosen endpoint, check if the given method exists, if not, choose the DEFAULT METHOD state.
+    const matchingMethod = Object.keys(
+      this.state[mostRelevantEndpoint],
+    ).includes(method)
+      ? method
+      : DEFAULT_STATE_HTTP_METHOD;
+    const relevantState = this.state[mostRelevantEndpoint][matchingMethod];
 
-    // Filter all the states that do not match the operation schema
-    const filteredStates = filterStatesByOperation(states, operation);
-    const parsedTopLevel = DSL.actTopLevelFromOAS(filteredStates);
-    // TODO: After parsing, do we want to see if we need to remove items from the current state?
-    return Object.keys(parsedTopLevel).length > 0 ? parsedTopLevel : undefined;
+    const { parsed, newState } = DSL.actTopLevelFromOAS(relevantState);
+    // Update state if needed
+    Object.keys(newState).forEach(code =>
+      Object.keys(newState[code]).forEach(media => {
+        const state = newState[code][media];
+        if (state === undefined) {
+          // Marked for deletion
+          this.deleteStateInternal(
+            mostRelevantEndpoint,
+            matchingMethod,
+            code,
+            media,
+          );
+        } else if (Object.keys(state).length > 0) {
+          // new state available
+          this.updateStateInternal(mostRelevantEndpoint, matchingMethod, {
```

Comment:
It's a bit surprising to me that a method called `getState` is doing all these updates and deletes to `this.state`, "get" sounds like something pure to me...

Reply:
I thought the same. At the same time, I don't think it's `Service`'s responsibility to do these updates.
My intention was that `getState` would actually indicate the return value, and what happens under the hood is abstracted away. Open for any suggestions though.

Reply:
At any case, the only time this update even happens (for now, at least), is when `$times` has reached 0, which is directly related to the `getState` operation. I don't see another reason why it would need to be updated after a get operation.

Reply:
Hmm, should there then be a simple `if (timesReachedZero(state)) { this.state = updateState(state); }` guard for that case?